storybookjs/storybook
# Use a custom `decorateStory` function in React

Issue: #12255 #11657, probably more

## What I did

To fix issue with stories remounting when re-rendering. The issue was the way the `decorateStory` function created a new "bound" story function each time you called it.

This meant React wasn't able to recognize the fact that we were just changing props to the story rather than rendering a whole new component.

The fix is to do the "binding" via React context rather than by creating a new bound function on each call.

Also, tweaked the loaders implementation so React could avoid rebinding the decorated story function with the loaded value each time.

## How to test

I added a story that renders "2" rather than the previous "4" value ;) Not sure if we have the test harness in place to check this via a proper react-rendered Jest test?

## NOTES:

 - I guess this won't work in React 15 (`createContext` didn't exist). What should we do here? I am leaning toward just using the default decorate story and not fixing the bug 🤷 

 - Type changes might be a bit funky, open to improvements.
